### PR TITLE
feat: add transaction bundler to relayer for better batching

### PR DESCRIPTION
**Pull request by mrice32**
**Motivation**

We'd like the relayer bot to do as much batching as possible.

**Summary**

Rather than doing the batching within the transaction-producing code, this adds a separate bundler that transactions are added to over the course of the bot run. Then, the bundler sorts the transactions by target and fires off batches that are as large as possible.

Note: this upgrades the typescript version to get access to the `Awaited` generic type. This caused some issues with compilation elsewhere in the repo, which were fixed in this PR.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [x]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

**Issue(s)**

N/A

### CODE REVIEW

**Review by ru0xjr669 — APPROVED** [review]

File: packages/insured-bridge-relayer/src/MulticallBundler.ts [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+  message?: string;
+  mrkdwn?: string;
+  level?: string;
+}
+
+type Call = BaseCall<TransactionType>;
+type ExtendedTransasction = TransactionType & { _parent: Contract };
+type ExtendedCall = BaseCall<ExtendedTransasction>;
+
+function isErrorSettlement<T>(input: PromiseSettledResult<T>): input is PromiseRejectedResult {
```
Comment (nicholaspai) [inline packages/insured-bridge-relayer/src/MulticallBundler.ts:25]: This function is also used in `index.ts`, maybe export it there and import here?

**Review by nicholaspai — APPROVED** [review]
Is this just a refactor of the existing bundling logic into a separate module, or is the logic for bundling different? If the latter, which situations are bundled differently now?

**Comment by mrice32** [comment]
> Is this just a refactor of the existing bundling logic into a separate module, or is the logic for bundling different? If the latter, which situations are bundled differently now?

@nicholaspai

Right now, the bundling happens per relayer object and per transaction type.

For example, let's say we're running the Optimism and Arbitrum bots in the same process and these four transactions were sent in the same run:
- WETH Relay from Optimism
- WETH Settlement for a relay sent from Optimism
- WETH Relay from Arbitrum
- WETH Settlement for relay sent from Arbitrum

Before this change, this would send 4 separate transactions. After this change, they would be bundled together.

File: packages/insured-bridge-relayer/src/MulticallBundler.ts (hunk 1 above)
Comment (mrice32) [reply packages/insured-bridge-relayer/src/MulticallBundler.ts:25]: I can make a helper file and import it.

Comment (mrice32) [reply packages/insured-bridge-relayer/src/MulticallBundler.ts]: Done.

File: packages/insured-bridge-relayer/src/index.ts [hunk 2]
```diff
@@ -170,14 +171,16 @@ export async function run(logger: winston.Logger, l1Web3: Web3): Promise<void> {
               if (config.botModes.l2FinalizerEnabled) await crossDomainFinalizer.checkForBridgeableL2TokensAndBridge();
               else logger.debug({ at: "AcrossRelayer#CrossDomainFinalizer", message: "L2->L1 finalizer disabled" });
 
+              // The multicall bundler likely accrued transactions over the course of the run.
+              // This call fires off those transactions, but does not wait on them to be mined.
+              // Note: we wait until this point to actually send off the transactions to
+              await multicallBundler.send();
```
Comment (nicholaspai) [inline packages/insured-bridge-relayer/src/index.ts:177]: Doesn't this batch transactions per relayer network? So All arbitrum relays + settles are batched, and on the next run, all optimism ...

**Review by nicholaspai — APPROVED** [review]
Just one comment in response to your general comment

File: packages/insured-bridge-relayer/src/index.ts (hunk 2 above)
Comment (mrice32) [reply packages/insured-bridge-relayer/src/index.ts:177]: You're totally right. I merged this incorrectly with the PR that added multichain support.

Comment (mrice32) [reply packages/insured-bridge-relayer/src/index.ts:177]: Fixed.

**Pull request merged**